### DISCUSSION TITLE
Obtain/extract snapserver's ip from snapclient machine

### DISCUSSION THREAD

**Discussion by jgabriel98** [snapcast/snapcast#733] [General]
## TLDR:
I want to know if there is someway to obtain the snapserver ip that the snapclient is connected, from the snapclient machine.

## Motivation behind:
I'am using snapcast with spotify connect, airplay and bluetooth support.
 - For spotify connect and airplay, wifi is the connection method, so where's a snapClient node, there is wifi nearby. <br/>Then i have playback control in all rooms (no problem).
 - but for Bluetooth, i have to be near the device (in this case, the machine with snapserver running).

So my idea was: instead of beeing tied to the snapserver machine's bluetooth, i could use the snapclient machines bluetooth too, configure alsa to output to a fifo file, and send the fifo file data back to snapserver through a tcp socket (since snapcast supports tcp), and then the snapserver would to his job of always (syncing).

I got this working already, but i had to hardcode the snapserver ip. But i don't want to make code specific for myself, i want to be a general solution.

#### My solution idea
So, i was thinking of extract the snapServer ip from the snapclient logs, like this:

the logs, from `$ journalctl -u snapclient`
<pre>
dez 02 16:03:18 raspberrypi-zero snapclient[2814]: (Browser) NEW: service 'Snapcast' of type '_snapcast._tcp' in domain 'local'
dez 02 16:03:18 raspberrypi-zero snapclient[2814]: (Browser) CACHE_EXHAUSTED
dez 02 16:03:18 raspberrypi-zero snapclient[2814]: Service 'Snapcast' of type '_snapcast._tcp' in domain 'local':
dez 02 16:03:18 raspberrypi-zero snapclient[2814]:         raspberrypi.local:1704 (192.168.1.107)
dez 02 16:03:18 raspberrypi-zero snapclient[2814]: Found server 192.168.1.107:1704
dez 02 16:03:18 raspberrypi-zero snapclient[2814]: Resolving host IP for: 192.168.1.107
dez 02 16:03:18 raspberrypi-zero snapclient[2814]: Connecting
dez 02 16:03:18 raspberrypi-zero snapclient[2814]: <b>Connected to 192.168.1.107</b>
dez 02 16:03:18 raspberrypi-zero snapclient[2814]: My MAC: "b8:27:eb:de:0a:11", socket: 8
dez 02 16:03:19 raspberrypi-zero snapclient[2814]: ServerSettings - buffer: 1000, latency: 0, volume: 100, muted: 0
dez 02 16:03:19 raspberrypi-zero snapclient[2814]: metadata:{"STREAM":"Music"}
</pre>

using this command: `$ journalctl -b -r -u snapclient.service | grep -m 1 -o -P "(?<=Connected to ).*"`
which would extract the ip from the last occurrence _Connected to \<server ip\>_

#### but...
This feels like an unstable workarround, so my question is: is there any other more robust way to find out which server the snapclient is connected to?

**Comment by jreusch** [comment]
hey hey
do not know if this solution is very ugly or just fits the needs, but i'd say it's much cleaner than parsing logs.

principal idea is parsing the ip out of the arguments with which a running snapclient process is running
```
$ cat /proc/$(pidof snapclient)/cmdline | sed 's/\o0/\n/g' | grep -A 1 -e "--host$" | tail -n 1
127.0.0.1
```

**Comment by jreusch** [comment]
beeing a bit more platform specific you could just expect the ip to which the snapclient connets to is configured in /etc/default/snapclient and get it out it, the same way as the systemd service does it
```
$ cat /lib/systemd/system/snapclient.service
[...]
EnvironmentFile=-/etc/default/snapclient
ExecStart=/usr/bin/snapclient --logsink=system $SNAPCLIENT_OPTS
```

**Comment by jgabriel98** [comment]
> hey hey
> do not know if this solution is very ugly or just fits the needs, but i'd say it's much cleaner than parsing logs.
> 
> principal idea is parsing the ip out of the arguments with which a running snapclient process is running
> 
> ```
> $ cat /proc/$(pidof snapclient)/cmdline | sed 's/\o0/\n/g' | grep -A 1 -e "--host$" | tail -n 1
> 127.0.0.1
> ```



> beeing a bit more platform specific you could just expect the ip to which the snapclient connets to is configured in /etc/default/snapclient and get it out it, the same way as the systemd service does it
> 
> ```
> $ cat /lib/systemd/system/snapclient.service
> [...]
> EnvironmentFile=-/etc/default/snapclient
> ExecStart=/usr/bin/snapclient --logsink=system $SNAPCLIENT_OPTS
> ```

okay, this make the "find out the server and send bluetooth data to it" step flexible, like i asked.
But it requires that the ip server is "hardcoded" on the snapclient config files. (gain flexibility at one part, loses on another)

My snapclient configuration(s) don't take the `--host` argument, it's all dynamic default. I know i could add this argument, but
i don't want to require any type of static ip or hardcoded ip.

**Comment by jreusch** [comment]
oh there is another method how a snapclient finds a snapserver? how? via avahi? if so i did not know of, so yes, this would then  not work, sorry :/

**Comment by jgabriel98** [comment]
I thinks is avahi but i'm not sure.
No problem, thank for trying to help anyway.

For now i'm sticking with the logs parsing. It's kinda dirty if if you think about it, but works very well on practice.

Until now i had no problems, and now i'm able to connect to any client's bluetooth.


### more thoughts:
Maybe one day we could open a feature request, where it would add a new capabilitie to snapclient:
 - send audio source data to snapserver.

this would be better than my general workarround:
<pre>
extract server ip ---_
                      \
                      \/
send audio data to server --> server send audio stream to clients
</pre>
because the snapclient already know who is the server, and could cache the audio data, so then the server woul just send timestamps without data to this client.

**Comment by jgabriel98** [comment]
If anyone is trying to do something similar, for curiosity or to save time, check my repo: [spotifypiHome](https://github.com/jgabriel98/spotifypiHome)
